File: Esame_28/JAVA/Gestione.java
import java.util.*;
import java.io.*;

public class Gestione {

	// LISTE
	static List<Veicolo> veicoli = new LinkedList<Veicolo>();
	static List<Posteggio> posteggi = new LinkedList<Posteggio>();
	// MAPPE
	static Map<Integer, Veicolo> codVeicolo = new HashMap<Integer, Veicolo>();

	public static void main(String[] args) {
		try {
			BufferedReader br = new BufferedReader(new FileReader("veicoli.txt"));
			String line = br.readLine(); // leggo la prima riga
			while (line != null) {
				StringTokenizer tok = new StringTokenizer(line);
				int codiceVeicolo = Integer.parseInt(tok.nextToken());
				String tipoVeicolo = tok.nextToken();
				String targaVeicolo = tok.nextToken();
				line = br.readLine(); // a capo
				if (tipoVeicolo.equals("auto")) {
					tok = new StringTokenizer(line);
					int cilindrata = Integer.parseInt(tok.nextToken());
					boolean diesel = Boolean.parseBoolean(tok.nextToken());
					line = br.readLine();// a capo
					String modelloVeicolo = line;
					line = br.readLine();// a capo
					String marcaVeicolo = line;
					line = br.readLine(); // a capo
					// Creazione nuovo oggetto
					Veicolo auto = new Auto(codiceVeicolo, tipoVeicolo, targaVeicolo, modelloVeicolo, marcaVeicolo,
							cilindrata, diesel);
					// aggiunta in lista
					veicoli.add(auto);
					// aggiunta in mappa
					codVeicolo.put(codiceVeicolo, auto);
				} else {
					tok = new StringTokenizer(line);// provo a vedere string eventualmente contenente spazi
					String categoria = tok.nextToken();
					line = br.readLine(); // a capo

					tok = new StringTokenizer(line);
					// visto che avevo: stringa eventualmente contenente spazi, allora:
					int numeroPosti = Integer.parseInt(tok.nextToken()); // problema del : for input string "9 " --> ci
																			// ho messo un tokenizer
					line = br.readLine(); // capo

					// tok = new StringTokenizer(line);
					String modelloVeicolo = line;// tok.nextToken();
					line = br.readLine();// a capo

					// tok = new StringTokenizer(line);
					String marcaVeicolo = line;
					line = br.readLine(); // a capo

					// Creazione nuovo oggetto
					Veicolo furgone = new Furgone(codiceVeicolo, tipoVeicolo, targaVeicolo, modelloVeicolo,
							marcaVeicolo, categoria, numeroPosti);
					// aggiunta in lista
					veicoli.add(furgone);
					// aggiunta in mappa
					codVeicolo.put(codiceVeicolo, furgone);
				}
			} // fine while
			br.close(); // chiusura file
		} // fine try
		catch (IOException e) {
			System.err.println(e);
		} catch (Exception e) {
			System.err.println(e);
		}

		try {
			BufferedReader br = new BufferedReader(new FileReader("posteggi.txt"));
			String line = br.readLine(); // leggo la prima riga
			while (line != null) {
				String nomeCognomeCliente = line;
				line = br.readLine();// a capo
				Posteggio post = new Posteggio(nomeCognomeCliente);
				// Gestione dell'elenco
				StringTokenizer tok = new StringTokenizer(line); // spezzo la linea in tokens
				while (tok.hasMoreTokens()) {
					int codiceVeicolo = Integer.parseInt(tok.nextToken());
					int numGiorni = Integer.parseInt(tok.nextToken());

					// Creazione elenco
					Elenco e = new Elenco(codVeicolo.get(codiceVeicolo), numGiorni); // gli passo i parametri così
					// aggiunta elenco in lista
					post.addElenco(e);
				}
				// Aggiunta oggetto in lista
				posteggi.add(post);

				line = br.readLine(); // a capo

			} // fine while
			br.close(); // chiusura file
		} // fine try
		catch (IOException e) {
			System.err.println(e);
		} catch (Exception e) {
			System.err.println(e);
		}

		// PUNTO 3:
		System.out.println("tipo, targa, codice, modello, marca, cilindrata, diesel, categoria, numero di posti");
		for (Veicolo v : veicoli)
			System.out.println(v);

		System.out.println();
		// PUNTO 4:
		for (Posteggio p : posteggi)
			System.out.println(p);

		System.out.println();
		// PUNTO 5:
		int codiceLetto = Integer.parseInt(args[0]); // codiceletto
		float somma = 0;
		for (Posteggio p : posteggi) {
			for (Elenco e : p.elenco) {
				if (e.veicolo == codVeicolo.get(codiceLetto))
					somma = somma + e.costo();
			}
		}
		System.out.println(codiceLetto + " " + somma);

	} // fine main

} // fine classe

File: Esame_28/PYTHON/Gestione.py
import sys
import argparse  # mi serve per fare il parsing

# IMPORTO LE CLASSI
from Veicolo import Veicolo
from Auto import Auto
from Furgone import Furgone
from Posteggio import Posteggio
from Elenco import Elenco


class Gestione:
    # COSTRUTTORE:
    def __init__(self, arg):
        self._arg = arg

    # METODO MAIN
    def main(self):
        veicoli = []  # lista 1
        posteggi = []  # lista 2
        codVeicolo = {}  # mappa o dizionario

        try:  # LETTURA PRIMO FILE
            f = open("veicoli.txt", "r")
            line = f.readline().strip()  # leggo la prima linea
            while line != "":
                tok = line.split()  # spezzo la linea
                codiceVeicolo = int(tok[0])
                tipoVeicolo = tok[1]
                targaVeicolo = tok[2]
                line = f.readline().strip()  # a capo
                if tipoVeicolo == "auto":
                    tok = line.split()  # spezzo la linea
                    cilindrata = int(tok[0])
                    diesel = bool(tok[1])
                    line = f.readline().strip()  # a capo
                    modelloVeicolo = line
                    line = f.readline().strip()  # a capo
                    marcaVeicolo = line
                    line = f.readline().strip()  # a capo
                    # CREAZIONE NUOVO OGGETTO
                    auto = Auto(codiceVeicolo, tipoVeicolo, targaVeicolo, modelloVeicolo, marcaVeicolo, cilindrata,
                                diesel)
                    # AGGIUNTA IN LISTA
                    veicoli.append(auto)
                    # AGGIUNTA IN MAPPA
                    # nomeDizionario[k] = V
                    codVeicolo[codiceVeicolo] = auto
                else:
                    tok = line.split()  # spezzo la linea
                    categoria = tok[0]
                    line = f.readline().strip()  # a capo

                    tok = line.split()  # spezzo la linea
                    # problema del 'for input string in Java'
                    numeroPosti = int(tok[0])
                    line = f.readline().strip()  # a capo

                    modelloVeicolo = line
                    line = f.readline().strip()  # a capo
                    marcaVeicolo = line
                    line = f.readline().strip()  # a capo

                    # CREAZIONE NUOVO OGGETTO
                    furgone = Furgone(codiceVeicolo, tipoVeicolo, targaVeicolo, modelloVeicolo, marcaVeicolo, categoria,
                                      numeroPosti)
                    # AGGIUNTA IN LISTA
                    veicoli.append(furgone)
                    # AGGIUNTA IN MAPPA
                    # nomeDizionario[K] = V
                    codVeicolo[codiceVeicolo] = furgone
                # fine if
            # fine while
            f.close()
        # fine try
        except IOError as i:
            print("IO Error found")
            print(i)
        except Exception as e:
            print("Unexpected error:", sys.exc_info()[0])

        try:  # LETTURA SECONDO FILE
            f = open("posteggi.txt", "r")
            line = f.readline().strip()  # leggo prima linea
            while line != "":
                nomeCognomeCliente = line
                line = f.readline().strip()  # a capo
                post = Posteggio(nomeCognomeCliente)
                # GESTIONE ELENCO(su una riga)
                tok = line.split()
                for i in range(len(tok) // 2):
                    codice = int(tok[i * 2])
                    numGiorni = int(tok[i * 2 + 1])
                    # nomeDizionario[K] --> ottengo il valore
                    e = Elenco(codVeicolo[codice], numGiorni)
                    # Agginta elenco in lista
                    post.addElenco(e)
                posteggi.append(post)
                line = f.readline().strip()  # a capo
            # fine while
            f.close()
        # fine try
        except IOError as i:
            print(i)

        # PUNTO 3:
        print("tipo, targa, codice, modello, marca, cilindrata, diesel, categoria, numero di posti")
        for v in veicoli:
            print(v.toString())

        # PUNTO 4:
        print("\n")
        for p in posteggi:
            print(p.toString())

        # PUNTO 5:
        try:
            codiceLetto = int(self._arg)  # codice letto
            somma = 0
            for post in posteggi:
                for e in post.getElenco():
                    if (e.veicolo == codVeicolo[codiceLetto]):
                        somma = somma + e.costo()
        except IOError as i:
            print(i)
        print(str(codiceLetto) + " " + str(somma))


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('veicolo', action='store')
    pa = parser.parse_args()
    g = Gestione(pa.veicolo)
    g.main()
